src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateObject = [];
+
+  let currentState = { ...state };
+
+  for (let i = 0; i < actions.length; i++) {
+    const { type, extraData, keysToRemove } = actions[i];
+    let nextState = { ...currentState };
+
+    switch (type) {
+      case 'clear':
+        nextState = {};
+        break;
+
+      case 'addProperties':
+        nextState = { ...nextState, ...extraData };
+        break;
+
+      case 'removeProperties':
+        for (let j = 0; j < keysToRemove.length; j++) {
+          delete nextState[keysToRemove[j]];
+        }
```

Comment:
For..of is also a better option here

```suggestion
        for (const keyToRemovle of keysToRemove) {
          delete nextState[keyToRemove];
        }
```